src/Streamly/Internal/Data/Stream/StreamD/Transform.hs

```diff
@@ -1203,28 +1203,28 @@ slicesBy p (Stream step1 state1) = Stream step (Just (state1, 0, 0))
 data RollingMapState s a = RollingMapInit s | RollingMapGo s a
 
 {-# INLINE rollingMapM #-}
-rollingMapM :: Monad m => (a -> a -> m b) -> Stream m a -> Stream m b
+rollingMapM :: Monad m => (a -> a -> m (Maybe a)) -> Stream m a -> Stream m (Maybe a)
```

Comment:
With this change rollingMapM just became useless for all other use cases, it is custom designed only for uniqBy use case. That's because, (1) the return type is restricted to be the same as the input type, (2)  the first element is always `Just`. Also the signature of rollingMap returns `a` whereas rollingMapM returns `Maybe a`, this is not expected.

Instead we can make the signature as:

```
rollingMapM :: Monad m => (Maybe a -> a -> m b) -> Stream m a -> Stream m b
```

The `Maybe` argument is the previous element, and the second argument is the current element. For the first element the previous element would be Nothing, for subsequent elements it would be Just.

Reply:
We can keep the original function as well, we can call it rollingMapM2 because it requires minimum 2 elements. It could be defined as:

```
rollingMapM2 :: Monad m => (a -> a -> m b) -> Stream m a -> Stream m b
rollingMapM2 f = catMaybes . rollingMapM g

  where

   g Nothing _ = Nothing
   g (Just x) y = Just (f x y)
```